### DISCUSSION TITLE
Grid_Daily for specific times?

### DISCUSSION THREAD

**Discussion by C00K50N** [StealthChesnut/HA-FoxESS-Modbus#53] [Q&A]
To anyone out there who's far more knowledgeable than me!...

I'd love to create a sensor which tells me how much kWh and £ I have imported from the grid so far today.  I know we have the grid_daily sensor, but since I'm on a time of use tariff (Octopus Go - cheaper rates between specified times overnight) I can't simply multiply that kWh figure.

So, can someone help me with how I'd go about creating a version of grid_daily but only for between say 00:30-04:30, and another for between 00:00-00:30 plus 04:30-23:59?  Then can I tie that to my £/kWh for those times to get an idea of cost so far that day?

I already have a "Helper" set up with the cost at those specified times, but I'm unsure how to tie that to the grid_daily.

Can anyone help me please?

**Comment by Daibutt** [comment]
Why not use the HA Energy dashboard? Enter  your helper in as the elec price and grid_total  (or whatever your import sum sensor is) as the meter and it should calculate it. I have Agile incoming & outgoing with 30min price changes. Once I got the prices into HA, the dashboard does the rest.
![image](https://user-images.githubusercontent.com/69276073/194371224-15faab7f-fd80-426a-aedc-8be79caacd4d.png)

Dashboard reports out like this, and will do day, week, month, year.
![image](https://user-images.githubusercontent.com/69276073/194371350-bc416346-88b9-4ba9-a4cb-3108330daf3e.png)

**Reply by C00K50N** [reply]
Thank you.  Yeah, that's a good shout.  I'm already set up that way, but the resulting figures aren't quite right.  Looking at it again I think I had put my tariff figures in without VAT so now I've corrected them it should come out correct on the Energy dashboard.

I'd still be keen to have it broken down into two, so I can see at a glance how effective I've been at shifting load onto the cheap rate.  And I tend to use my own custom dashboard for things at a glance, so it would be nice to be able to have those figures to hand there.

Incidentally, your Agile figures make good reading!  We don't export much but use a lot, so I don't know if we could make that worthwhile, but maybe it's something to consider next summer.

**Reply by Daibutt** [reply]
Ah OK, if you want to split it out like that you would probably need to create seperate templates for your incoming power. Makes it a bit more messy tho.....

I'm not going to try and YAML it on the hoof but in essence;

Sensor: Daytime power
If time > 04:30 and < 23:59
value = incoming power 
else value = 0

Sensor : Cheap power
If time > 00:00 and < 04:29
value = incoming power
else value = 0

Set up seperate integration sensors for each, that will give you your kWh for the different tariff periods
i.e. Daytime power sum & Cheap power sum

Then add more (!) template sensors for the cost in each period;

Daytime power cost = Daytime power sum X daytime power price
Cheap power cost = Cheap power sum X cheap power price

Or something like that!!


Oh and as for the Agile numbers..... wind has picked up in the last week so that's why they look OK, as import prices are low at times and I have decent amount of storage to tide me over when they are not.  Roll on the wind-less days in December when I'm generating FA solar and paying 34p to import lol.
Been on Agile in/out for over a year now. 
In 12 months my NET import cost was about £16 + standing charge £77, + VAT on imports of ~ £20  (even though I'm a net exporter I pay VAT on imports in winter but obv no VAT on receipts for export in summer).
I don't have ASHP or an EV so that makes a massive difference & I probably wouldn't be on Agile if I did !!

**Reply by C00K50N** [reply]
Sounds great!  Yeah, we have ASHP and an EV, so we use more electricity than most!  Solar install in July, now including batteries which arrived a couple weeks ago is already helping a lot and should mean this winter is a lot less expensive than the last couple years!

I'm loving the HA integration, as FoxESS's own online portal is pretty poor.  And I like tinkering!

Many thanks for your suggestions above.  I've had a play with it and I feel like I'm getting somewhere, but my knowledge of coding tends to let me down.  I usually try to copy / paste similar things until I can get them to work, but I've run out of luck here.

I've managed to get an "if / or" line working for the peak times of 00:00-00:30 and 04:30 onwards, but when I try and use the sensor.grid_consumption it seems to return the current grid kW rather than the accumulated kWh within the specified period.  I suspect I may have created the wrong type of sensor perhaps?  I tried asking it to return the sensor.grid_consumption_sum but that returns the total so far, and sensor.grid_daily which returns the daily total of peak and off-peak.

Here's my code so far...

    grid_daily_peak:
          friendly_name: "Peak Rate kWh Today"
          unit_of_measurement: 'kWh'
          device_class: energy
          value_template: >
            {% if (now().strftime('%r') < '00:30 AM') or (now().strftime('%r') > '04:30 AM') %}
            {{states('sensor.grid_consumption')}}
            {% else %}
            0
            {% endif %}

I would welcome any further suggestions for the first one, then hopefully I can work out the rest from there.... hopefully!

**Reply by Daibutt** [reply]
grid_consumption is your instant power reading I think (kW) not energy (kWh). At least it is in mine... if you put kettle on grid_consumption should increase by 2 or 3 kw then drop back when kettle turns off?

If so, you'll need to use the same sort of if/then logic as above to create two power sensors (kW ..... not kWh which is energy). One will show the same as grid consumption  00:30> 04:30, and zero outside this time, the other will do the opposite (04:30-00:30).

Once you have these you will need to add two new integration sensors similar to ;

- platform: integration
  name: grid_consumption_sum_cheap (or peak)
  source: sensor.grid_consumption cheap (or peak)
  unit_time: h
  round: 2

But one for each of your new power sesnsors (cheap & daytime) - these will do the conversion from kW power to kWh energy.
Now you should be able to add these as 2 separate grid consumption sensors to the HA energy dashboard and put the appropraite helper price sensor in each one (or a fixed number if they stay the same) & the HA dashboard should do all the hard work.
![image](https://user-images.githubusercontent.com/69276073/194697424-269765f3-86e8-4a43-b425-c8d314ad209f.png)




If that doesn't work or you don't want to use the HA dashboard then it gets more fiddly;



These totalisers will just increase indefinitely (like a mileometer in a car) so to get your daily energy kWh you'll need to add utility meter sensors (bit like a trip meter in a car) for the two new integration sensors, similar to this;

utility_meter:
  grid_daily:
    source: sensor.grid_consumption_sum_cheap (or peak)
    cycle: daily

Now these 2 new meter sensors will give you your daily total kWh for cheap and day rate periods.
To get your costs add some template sensors;

Daily cost cheap = cheap rate daily utility meter (kWh) * cheapr tariff rate (p/kWh)
Daily cost peak = peak rate daily utility meter (kWh) * peak tariff rate (p/kWh)

**Reply by C00K50N** [reply]
That’s fantastic! Thanks so much for your detailed reply. I look forward to trying all that soon. I’ll report back to let you know how I get on.

**Comment by Daibutt** [comment]
This is essentially what you are trying to create if you go the HA dashboard route;


![image](https://user-images.githubusercontent.com/69276073/194712348-ffd00d72-8099-412f-a2bc-9aff22e79bd6.png)

**Reply by Daibutt** [reply]
Oh, I meant to say, you've done the hard bit by getting the if/then time stuff working 👍 
Once you set up separate sensors showing current peak/cheap power usage, you just replicate the existing integration and utilty sensors for each and feed them into the dashboard.
Let me know how you get on. I don't have dual tariffs like you but definitely interested in seeing this work. It should do.

**Reply by C00K50N** [reply]
That really is brilliant!  Thank you again.  I've had a mad busy weekend so didn't get a chance to look at it again, but I'll try and find time today.  I'll definitely report back.

**Reply by andyhasalandy** [reply]
Watching this thread with keen interest! Perhaps a little lazy, but do share once you have working config :-)

**Reply by C00K50N** [reply]
Working on it now, and so far so good it seems...  Or at least I don't have any errors when I check the configuration!!

Just at the stage of adding 2 template sensors to get the daily costs for each rate.  I'm not sure if I'm doing it right, but I'll see how we go!  Hopefully not long until I report back...

**Reply by andyhasalandy** [reply]
Fantastic- I current use this helper as a way of capturing a total daily cost, sharing in case it is of use : 
https://ryanteck.uk/octopus-go-hassio/amp/

**Reply by C00K50N** [reply]
Right...  I'm hoping that's all set up, but I'm not sure!  It's not easy to really test it just now since we're not using any grid power this evening.  So it's all zeros or unknown for now, but I hope that will change overnight when we've charged the batteries from the grid.

Here's an output I've quickly set up for now:
![image](https://user-images.githubusercontent.com/102669305/194938825-0003130f-1a30-48d1-8cd3-b104fa911fc8.png)

And below is the code of how I got there (code for peak shown, but for each I have another for cheap as well).  Note the later times since our previous Go tariff only ran until yesterday so we moved to Go Faster today...

Template sensors to read power (kW) only at the specified times:

    grid_consumption_peak:
          friendly_name: "Peak Rate Consumption"
          unit_of_measurement: 'kW'
          device_class: power
          value_template: >
            {% if (now().strftime('%r') < '01:30 AM') or (now().strftime('%r') > '06:30 AM') %}
            {{states('sensor.grid_consumption')}}
            {% else %}
            0
            {% endif %}

Integration sensors to get energy (kWh) from above power values:

- platform: integration
  name: grid_consumption_sum_peak
  source: sensor.grid_consumption_peak
  unit_time: h
  method: left
  round: 2

Utility Meters to reset energy totals daily:

  grid_daily_peak:
    source: sensor.grid_consumption_sum_peak
    cycle: daily

Template sensors to show cost so far that day:

    daily_cost_peak:
          friendly_name: "Higher Rate Cost Today"
          unit_of_measurement: '£'
          value_template: >
            {{states('sensor.grid_consumption_sum_peak') * 0.3882}}

I'm not sure if I've done them right, particularly the last one.  There may very well be some glaring errors in my code!  I'll happily be corrected, but it feels like we're heading in the right direction.

**Reply by C00K50N** [reply]
Yes!!  I think we're getting somewhere...
![image](https://user-images.githubusercontent.com/102669305/194942169-d6640a2f-4838-4bf8-9e60-d9a2f4c6f068.png)

**Reply by Daibutt** [reply]
Nice work! 👍

**Comment by C00K50N** [comment]
Hmm... I've obviously missed something.  This is how my list of outputs looks this morning...
![image](https://user-images.githubusercontent.com/102669305/195018244-a0895ea8-1b8d-4ad0-b696-0861bee9dd2c.png)

The grid_daily_ counters look correct - the two of them add up to the original grid_daily one.  But I must have done something wrong with the cost ones.  Here's my code now - most of which I have to admit I guessed at based on bits and pieces I found online!  Can anyone suggest where I've gone wrong?

    daily_cost_peak:
          friendly_name: "Higher Rate Cost Today"
          unit_of_measurement: '£'
          icon_template: mdi:currency-gbp
          value_template: >
            {{'{:,.2f}'.format(float(states('sensor.grid_consumption_sum_peak')|float * 0.3882 ))}}

**Reply by Daibutt** [reply]
You have a sensor called grid_daily_peak in the card but the code uses sensor.grid_consumption_sum_peak, is this intentional? Not sure that is the cause mind you!

The HA dashboard will do the math for you if you add these two grid_daily kWh counters in with their respective prices. 
You can then still access those daily costs outside of the dashboard if you want them in other cards/displays.
Search for sensors with "monetary" attributes, my elec import cost one (that the dashboard calculates) is called sensor.daily_electric_import_cost, your two will be different obviously.
![image](https://user-images.githubusercontent.com/69276073/195022408-fde6ecce-ea7f-4ceb-b8e3-cb71f5ff15b8.png)

You can also do this for gas and elec export if configured.
I don't do any of the kWh * GBP/kWh calcs in my config, but just pull the HA dashboard generated numbers.
Only minor issue is they reset to zero if you reboot HA (but weirdly not in the dahsboard itself!?) but not a real problem unless you do a lot of HA reboots.

**Reply by C00K50N** [reply]
Ah, nice.  Thank you, yeah looks like the built-in ones have added up correctly, so I should be able to use them in my own list on my other dashboard.  Lovely!
![image](https://user-images.githubusercontent.com/102669305/195028590-63fb673b-807a-491c-b080-1e95af1005b3.png)

Thank you for spotting the sensor too - that might indeed be where I've gone wrong in my own code.  I'll try again later.

**Comment by C00K50N** [comment]
Many thanks to @Daibutt for all your help with setting up the cheap and peak sensors.  I now have my dashboard set up pretty much how I'd like it, with all the information I want to see at a glance...
![image](https://user-images.githubusercontent.com/102669305/195293796-059121f8-ef6f-4c48-9d0f-c1d462dcf91a.png)
(viewing it on my phone I see all the gauges and info on the right first, then the statistics graphs lower down, which works well since the graphs don't read very well on the phone)

In case it's helpful for others, here's all the added code, like I sent previously in this thread but with some tweaks and corrections...

Template sensors to read power (kW) only at the specified times:

    grid_consumption_peak:
          friendly_name: "Import Now Higher Rate"
          unit_of_measurement: 'kW'
          device_class: power
          value_template: >
            {% if (now().strftime('%r') < '01:30 AM') or (now().strftime('%r') > '06:30 AM') %}
            {{states('sensor.grid_consumption')}}
            {% else %}
            0
            {% endif %}

    grid_consumption_cheap:
          friendly_name: "Import Now Cheap Rate"
          unit_of_measurement: 'kW'
          device_class: power
          value_template: >
            {% if (now().strftime('%r') > '01:30 AM') and (now().strftime('%r') < '06:30 AM') %}
            {{states('sensor.grid_consumption')}}
            {% else %}
            0
            {% endif %}

Integration sensors to get energy (kWh) from above power values:

- platform: integration
  name: grid_consumption_sum_peak
  source: sensor.grid_consumption_peak
  unit_time: h
  method: left
  round: 2
- platform: integration
  name: grid_consumption_sum_cheap
  source: sensor.grid_consumption_cheap
  unit_time: h
  method: left
  round: 2

Utility Meters to reset energy totals daily:

  grid_daily_peak:
    source: sensor.grid_consumption_sum_peak
    cycle: daily
  grid_daily_cheap:
    source: sensor.grid_consumption_sum_cheap
    cycle: daily

And Template sensors to show cost so far that day:

    daily_cost_peak:
          friendly_name: "Cost Today Higher Rate"
          unit_of_measurement: '£'
          icon_template: mdi:currency-gbp
          value_template: >
            {{'{:,.2f}'.format(float(states('sensor.grid_daily_peak')|float * 0.3882 ))}}

    daily_cost_cheap:
          friendly_name: "Cost Today Cheap Rate"
          unit_of_measurement: '£'
          icon_template: mdi:currency-gbp
          value_template: >
            {{'{:,.2f}'.format(float(states('sensor.grid_daily_cheap')|float * 0.0825 ))}}

As far as I can see it's working well!  Some of my figures seem to be ever so slightly different to those on the built-in Energy dashboard. 
 It's only by a small amount so I wonder if it's related to how often sensors poll or something like that.  Regardless, it's not an issue for me; figures from both are also slightly different to those on my energy bills, but everything is close enough that I'm OK with it.

I'll call this a success; thank you again @Daibutt for your help.

**Reply by Daibutt** [reply]
Nice job! Glad you got it working 👍

**Reply by andyhasalandy** [reply]
Thank you. Config added!

**Reply by andyhasalandy** [reply]
Looking at the stats after a couple of days, I’m not sure we’ve got the filtering criteria right yet on the sensors. I have peak consumption (in the day) landing into my off peak (12 hours offset from my off peak window).  Also the peak sensor seems to keep counting until 01:00 rather than 00:30 - quite noticeable when the EV kicks in at 00:30.
Just me?

**Reply by C00K50N** [reply]
Our amazing Go tariff (00:30-04:30 cheap) ran out while I was in the process of working this out.  So while my first figures were for those times, my final code reflects our new Go Faster tariff (01:30-06:30 cheap).  Forgive me if you've already spotted this, but if you've copied the code without changing the times on the if/then lines, perhaps that's where the discrepancy comes in for you.  Mine seems to be working spot on.

If your tariff has your cheap rate at different times, you'll want to find the _grid_consumption_peak_ and _grid_consumption_cheap_ template sensors, and change your times on both accordingly.

I'm not sure why it would be reading as cheap during the day though, 12 hours offset.  Mine isn't doing that as far as I know.  It's difficult to be sure to be fair, as we are so far managing to use little or no grid import during the day as we're working just from the batteries most days.

**Comment by andyhasalandy** [comment]
Noted, cheers.  I’d updated those already, I think it relates to the time format code returning only the hour rather than the full ‘24h’ format.  I’ve updated to use %H and 00:30:00 etc…. And will see how it behaves overnight!

**Reply by C00K50N** [reply]
Ah, nice.  Yeah, I'm not sure why mine wasn't in 24hr format, but I seemed to need AM and PM.

I hope you can get it to work too.
